cmd/kube-proxy/app/init_windows.go

```diff
@@ -0,0 +1,54 @@
+// +build windows
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"os"
+
+	"k8s.io/kubernetes/pkg/windows/service"
+
+	"github.com/golang/glog"
+	"github.com/spf13/pflag"
+)
+
+const (
+	serviceName        = "kube-proxy"
+	serviceDisplayName = "Kubernetes network proxy"
+)
+
+func initForOS() {
+	// On Windows, this may be launching as a service or with an option to
+	// register the service.
+	stop, runAsService, err := service.InitService()
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	if stop {
+		os.Exit(0)
+	}
+
+	if runAsService {
+		glog.Infof("Running as service!")
+	}
+}
+
+func addOSFlags(fs *pflag.FlagSet) {
```

Comment:
@ncdc regarding kube-proxy flag registrations

Reply:
Shouldn't these go into config instead of being flags?

Reply:
I haven't figured out how I want to represent OS-specific configuration parameters in the component config structs yet, so for Kubelet I'd like to keep these as flags for now...

kube-proxy is maybe another story, since we tried to avoid the flag-precedence magic the kubelet implements that allows flags to seamlessly graduate into config fields...

Reply:
In my opinion this would work only as flags. All the flags are being used only to detect if the user wants to register/remove the kube-proxy as a Windows service.
The flags ```--register-service```/```--unregister-service``` and others added by this should only be called the first time. When the kube-proxy is registered as a Windows service it will not have those flags passed to it, as these are required only the first time to do the Windows registration.
If you would still really like to have those inside the config, I will try to find a way to fit them there.

Reply:
I'm ok with these being flags. They're similar to kube-proxy's `--cleanup-iptables` in that they're one-shot immediate directives.

Reply:
>should only be called the first time

This feels like a strange semantic. You should be able to say "these are the args my thing should run with," and the same set should work for the initial and all subsequent invocations without change - especially for something like kube-proxy, which is specified via a Pod.

As a general recommendation, try to design declarative APIs rather than imperative ones. 

Edit: I guess what you're trying to do is use kube-proxy or kubelet as a bootstrapper for kube-proxy/kubelet as windows services... Is this really what we want to do? 

In kube-proxy's case, don't we lose the visibility we'd typically have via the Pod - not to mention start making lifecycle policy decisions outside of the Kubernetes API?

Kubelet doesn't decide to set itself up as a systemd service, so why should it do what sounds like the same thing for windows? (I know nothing about windows, FYI).

Reply:
@mtaufen I'm not sure if I follow your logic, but I will try to add more context on what we are trying to achieve.

To break it in more simpler steps `--run-service` is required for the binary to run as a Windows service. Think of having a systemd service but you can only run as a systemd service if and only if you subscribe via an API.

`--register-service`/`--unregister-service` is equivalent to creating/deleting your service via sc.exe (https://support.microsoft.com/ro-ro/help/251192/how-to-create-a-windows-service-by-using-sc-exe for more info) with the caveat that is a commodity for the user. Windows services have a lot of options that can be set: dependencies, restart options with different counters, triggers via events etc. Usually creating/deleting services is done via MSI's, but due to the lack of a Kubernetes MSI we opted to bottle up the experience of creating/deleting the services via those two command line arguments.
`Kubelet doesn't decide to set itself up as a systemd service, so why should it do what sounds like the same thing for windows? `
Nor should it. The user is the one deciding to run the kubelet as a service because the user is the one that passes the command line arguments. Also this is the same approach that has been done on Docker for Windows.
Does the above make sense?

Reply:
>Think of having a systemd service but you can only run as a systemd service if and only if you subscribe via an API.

Is it a hard requirement on Windows that when run as a service a binary must also self-identify as a service? If so, why can't we achieve this with a single `--service` flag?

> Usually creating/deleting services is done via MSI's, but due to the lack of a Kubernetes MSI we opted to bottle up the experience of creating/deleting the services via those two command line arguments.

Ideally, the decision of whether or not to run something as a service would be completely decoupled from the binary and its command-line. If adding these args is just a shortcut because we don't have an MSI, that makes it sound like the right solution is to create an MSI.

> The user is the one deciding to run the kubelet as a service because the user is the one that passes the command line arguments.

I'm not sure I have a clear understanding of the user story here. What are the reasons for running something as a service, and in which cases are these already covered by existing Kubernetes APIs? E.g. I could imagine you want it to be automatically restarted if it crashes, but in the kube-proxy case, why not configure it as a Pod with RestartPolicy=Always?

Reply:
`Is it a hard requirement on Windows that when run as a service a binary must also self-identify as a service?`
It is a hard requirement. We could use wrappers i.e: https://nssm.cc/usage or https://github.com/cloudbase/OpenStackService, but you gain more advantages when doing from the actual binary.
`If so, why can't we achieve this with a single --service flag?`
We could. Although we prefer possible to have all three options for the moment. We shall remove this step when we can bundle the kubernetes installation via MSI.
`I'm not sure I have a clear understanding of the user story here. What are the reasons for running something as a service, and in which cases are these already covered by existing Kubernetes APIs? E.g. I could imagine you want it to be automatically restarted if it crashes, but in the kube-proxy case, why not configure it as a Pod with RestartPolicy=Always?`
That would be just one feature that you can access.
There are lot of hidden dragons that you can unlock. I.E. Running as an elevated user i.e. LocalSytstem (https://msdn.microsoft.com/en-us/library/windows/desktop/ms684190(v=vs.85).aspx).
You can add triggers for first/second/rest of the restarts to trigger a script(cleanup/dumps/etc) and reboot the node if required.
Instead of starting blindly you can wait for dependencies: i.e. I don't have docker or kubelet running, I add docker as a dependency to the kubelet service. Whenever I start the kubelet service I also start docker. Service manager is the one that starts docker and after starts kubelet. 
I could add more examples if needed.

Reply:
>It is a hard requirement.

Am I understanding right that even with the MSI we'll still need something like `--service`? In that case, ok then. Can't fight the OS APIs. 

>We shall remove this step when we can bundle the kubernetes installation via MSI.

How long will it take to create an MSI/what is the risk of *not* having these flags in this release? I'd rather not even have to carry the temporary bits if possible.

Reply:
I don't think we should count on MSI as a means for configuration management. It hasn't received any new features in years. The guidance for desktop developers since Windows 8 has been to move to APPX but there's no equivalent for servers. This is the same reasoning that there isn't a MSI installer for Docker EE-basic on Windows Server.

I think this solution with the `--register-service` flag is the right one and will be easier to use with configuration managers such as PowerShell Desired State Configuration, Ansible, Chef, Puppet, and so on. MSI is not idempotent, and that's a property that configuration managers prefer to have.

Reply:
>MSI is not idempotent, and that's a property that configuration managers prefer to have.

These are not idempotent flags, because they have to be stripped from the command line before registering the component as a service.

>use with configuration managers such as PowerShell Desired State Configuration, Ansible, Chef, Puppet, and so on

are *none* of these capable of configuring something to run as a service?

Reply:
They could shell out to `sc.exe` to register a service. 

My main point was if you ran a playbook that called `kubelet.exe --register-service ...` it would do the right thing. If you modified your playbook with different flags after that, the new settings from the next `kubelet.exe --register-service` would still apply correctly. `--unregister-service` would always work as well. 

If you don't want functionality that makes it easy to configure the service, it could be removed and handled as extra steps in an outside system but `--service` is absolutely essential to enable the Windows service control manager API integration.

Reply:
`Am I understanding right that even with the MSI we'll still need something like --service? In that case, ok then. Can't fight the OS APIs.`
Yup `--service` is mandatory. It's OS specific.
`How long will it take to create an MSI/what is the risk of not having these flags in this release? I'd rather not even have to carry the temporary bits if possible.`
I think the next release cycle might have support we might add support for it (if we find the time for it).
@mtaufen the main risk here if we drop `--register-service`/`--unregister-service` is that we have to at least document it somewhere, but I'm not exactly sure where and if the documentation changes can land into 1.10 milestone.
How should we proceed?

Reply:
Can you point me to links where documentation would need to change?
We didn't have `--register-service`, etc. before this PR, so I'm not sure why there would be documentation indicating otherwise. 
Generally, documentation lives in https://github.com/kubernetes/website.

Regarding `--service`, is there any way we could have the Windows build of a Kubelet automatically detect that someone is trying to configure it as a service, and avoid needing any flags at all?

Reply:
@mtaufen unfortunately `--service` is `mandatory`.

We are dropping the `--register-service`/`--unregister-service` part from this PR.

Reply:
>unfortunately --service is mandatory

Oh well. Ok then, that'll have to do.

>We are dropping the --register-service/--unregister-service part from this PR.

I'm confused. Do existing docs need to be updated to reflect this, or did you just want to add documentation that clearly indicates what the Kubelet is/isn't capable of wrt Windows SCM?